tt_metal/impl/device/multi_device.cpp

```diff
@@ -23,43 +22,46 @@ DeviceMesh::DeviceMesh(const DeviceGrid& device_grid, const DeviceIds &device_id
     TT_ASSERT(num_requested_devices <= num_available_devices, "Requested more devices than available");
     TT_ASSERT(num_requested_devices <= device_ids.size(), "User provided insufficient number of device_ids for DeviceMesh");
 
-
-    //TODO: for DevicePool feature delete CreateDevices and merge with this function
-    //TODO: should there be an explicit CloseDevices call somewhere?
     bool is_galaxy = tt::Cluster::instance().is_galaxy_cluster();
-    std::vector<chip_id_t> mmio_device_ids = {};
-    if (is_galaxy) {
-        mmio_device_ids.push_back(0);
-        if (num_requested_devices > 8) {
-            mmio_device_ids.push_back(1);
-        }
-        if (num_requested_devices > 16) {
-            mmio_device_ids.push_back(2);
-        }
-        if (num_requested_devices > 24) {
-            mmio_device_ids.push_back(3);
-        }
-    } else {
-        mmio_device_ids = device_ids;
-    }
-    managed_devices = tt::tt_metal::detail::CreateDevices(mmio_device_ids, 1, l1_small_size, trace_region_size);
     if (is_galaxy) {
+        // Temp solution until we add algorithmic way to determine chip connectivity
+        // Map col to tunnel depth and row to tunnel count
+        int cluster_tunnel_depth = tt::Cluster::instance().get_mmio_device_max_tunnel_depth(0);
+        int cluster_tunnel_count = tt::Cluster::instance().get_mmio_device_tunnel_count(0);
+        int num_mmio_devices = tt::Cluster::instance().number_of_pci_devices();
+        TT_ASSERT(num_cols <= cluster_tunnel_depth and num_rows <= cluster_tunnel_count * num_mmio_devices, "Unsupported Galaxy mesh shape");
+
         DeviceIds galaxy_device_ids;
-        for (const auto &[dev_id, dev]: managed_devices) {
-            galaxy_device_ids.emplace_back(dev_id);
+        for (int mmio_device_id = 0; mmio_device_id < num_mmio_devices; mmio_device_id++) {
```

Comment:
This assumes mmio_device ids are zero-indexed?

Reply:
yes

Reply:
so is there a way to just fetch the mmio-device ids directly and iterate over those?

Reply:
I'll look around, I think there is an api for that. However this kind of looping is actually more in line with what syseng and rest of runtime is doing..

Reply:
I actually don't think that would be correct. If the mmio device ids are returned as like (0,2,1,3) from some other api, that wouldn't yield the right order for the rows of chips.

Reply:
I'm assuming the method that returns mmio device ids is sorted on rows... We're implicitly assuming mmio_device_index == mmio_device_id in sorted order. This would just be wrong if the actual correct ordering was (0,2,1,3), right?

Reply:
We have no clean of properly returning (0,2,1,3) in the right order until we build something that checks the eth connectivity.

Reply:
okay, let's defer for later then